kotlinx-coroutines-core/jvm/src/ThreadContextElement.kt

```diff
@@ -111,12 +111,21 @@ public interface ThreadContextElement<S> : CoroutineContext.Element {
  *         traceThreadLocal.set(oldState)
  *     }
  *
- *     override fun copyForChildCoroutine(): CopyableThreadContextElement<TraceData?> {
+ *     override fun copyForChildCoroutine(): TraceContextElement {
  *         // Copy from the ThreadLocal source of truth at child coroutine launch time. This makes
  *         // ThreadLocal writes between resumption of the parent coroutine and the launch of the
  *         // child coroutine visible to the child.
  *         return TraceContextElement(traceThreadLocal.get()?.copy())
  *     }
+ *
+ *     override fun merge(element: TraceContextElement): TraceContextElement {
```

Comment:
>Is that roughly correct?

Yes. Context elements polymorphism is a mine field that require a discipline to follow, otherwise the element may start to interact unpredictably, depending on how `minusKey` and `plus` are invoked on the context.

>That can happen sometimes for reasons I haven't fully understood. 

That's likely to be `Dispatchers.Unconfined` and `CoroutineStart.UNDISPATCHED`

Reply:
> Yes. 

OK.

Then I think there are two considerations:

1. Which is the more error-prone API: `copy()` + `merge()` (though it seems `copy()` alone might be OK?), or a delegated `newCoroutineContext()`?
2. Which would Coroutines prefer to support and have to avoid accidentally breaking with future changes?

Either API seems sufficient, and I don't think a third option will soon present itself.

> Context elements polymorphism is a mine field that require a discipline to follow, otherwise the element may start to interact unpredictably, depending on how minusKey and plus are invoked on the context.

Agreed.

> That's likely to be Dispatchers.Unconfined and CoroutineStart.UNDISPATCHED

I had a suspicion. :)